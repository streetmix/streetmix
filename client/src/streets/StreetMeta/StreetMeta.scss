@use "sass:color";
@import "../../../styles/variables";

.street-meta {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  padding-top: 0.5em;
  color: black;
  user-select: none;
  pointer-events: auto;
  transition: color $skybox-transition;
  font-size: 0.925rem;

  > div:not(:first-child) {
    margin-left: 0.25em;
  }

  /* Invert the UI text color when there is a dark background */
  body.dark-skybox-invert-ui & {
    color: #fff;
  }
}
