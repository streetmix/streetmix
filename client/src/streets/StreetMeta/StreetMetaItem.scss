@import "../../../styles/variables";

.street-meta-item {
  display: flex;
  align-items: center;
  white-space: nowrap;
  margin-bottom: 0.2em;

  button {
    display: flex;
    align-items: center;
    justify-content: center;
    height: 2rem;
    padding: 0 0.5em;
    cursor: pointer;
    border-radius: $border-radius-large;

    &:hover {
      background-color: rgba(255 255 255 / 20%);
    }
  }

  // Links that appear in meta items (currently: only in Author)
  // should have the same text color as rest of component
  a {
    color: inherit;

    // In read-only mode, links are still interactive, but
    // made to look like other text
    body.read-only & {
      text-decoration: none;
      cursor: default;
    }
  }
}

.street-meta-item-plain {
  // Match button dimensions but not behavior
  display: flex;
  align-items: center;
  justify-content: center;
  height: 2rem;
  padding: 0 0.5em;

  // Underlines are hidden
  .underline {
    text-decoration: none;
  }
}

.street-meta-icon {
  display: flex;
  align-items: center;
  justify-content: center;
  margin-right: 0.3em;

  svg {
    width: 16px;
    height: 16px;
  }
}

.street-meta-content {
  display: flex;
  flex-direction: row;
  align-items: center;
}
