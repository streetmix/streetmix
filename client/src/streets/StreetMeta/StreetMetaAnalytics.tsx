import React from 'react'
import { FormattedMessage, useIntl } from 'react-intl'

import { useSelector, useDispatch } from '~/src/store/hooks'
import { showDialog } from '~/src/store/slices/dialogs'
import { getStreetCapacity } from '~/src/segments/capacity'
import Icon from '~/src/ui/Icon'
import { formatNumber } from '~/src/util/number_format'
import StreetMetaItem from './StreetMetaItem'

function StreetMetaAnalytics (): React.ReactElement | null {
  const street = useSelector((state) => state.street)
  const locale = useSelector((state) => state.locale.locale)
  const dispatch = useDispatch()
  const intl = useIntl()

  const averageCapacity = getStreetCapacity(street).average ?? 0
  const tooltip = intl.formatMessage({
    id: 'dialogs.analytics.heading',
    defaultMessage: 'Analytics'
  })

  // If zero capacity, don't display anything
  if (averageCapacity > 0) {
    return (
      <StreetMetaItem
        isEditable={true}
        tooltip={tooltip}
        onClick={() => dispatch(showDialog('ANALYTICS'))}
        icon={<Icon name="graph" />}
      >
        <span className="underline">
          <FormattedMessage
            id="capacity.ppl-per-hour"
            defaultMessage="{capacity} people/hr"
            values={{ capacity: formatNumber(averageCapacity, locale) }}
          />
        </span>
      </StreetMetaItem>
    )
  }

  return null
}

export default StreetMetaAnalytics
