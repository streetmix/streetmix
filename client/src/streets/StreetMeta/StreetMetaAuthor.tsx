import React from 'react'
import { FormattedMessage } from 'react-intl'
import { IoBrushOutline } from 'react-icons/io5'
import { useSelector, useDispatch } from '../../store/hooks'
import { openGallery } from '../../store/actions/gallery'
import { useGetUserQuery } from '../../store/services/api'
import Avatar from '../../users/Avatar'
import { isOwnedByCurrentUser } from '../owner'
import StreetMetaItem from './StreetMetaItem'

function StreetMetaAuthor (): React.ReactElement | null {
  const creatorId = useSelector((state) => state.street.creatorId)
  const signedIn = useSelector((state) => state.user.signedIn)
  const userId = useSelector((state) => state.user.signInData?.userId ?? '')
  const { data: creatorProfile } = useGetUserQuery(creatorId)
  const dispatch = useDispatch()

  function handleClickAuthor (event: React.MouseEvent): void {
    event.preventDefault()
    void dispatch(openGallery({ userId: creatorId }))
  }

  let user
  if (creatorId !== null && (!signedIn || creatorId !== userId)) {
    user = (
      <>
        <Avatar userId={creatorId} />
        <a href={'/' + creatorId} onClick={handleClickAuthor}>
          {creatorProfile?.displayName ?? creatorId}
        </a>
      </>
    )
  } else if (creatorId === null && (signedIn || !isOwnedByCurrentUser())) {
    user = <FormattedMessage id="users.anonymous" defaultMessage="Anonymous" />
  }

  if (!user) return null

  return (
    <StreetMetaItem icon={<IoBrushOutline />}>
      <FormattedMessage
        id="users.byline"
        defaultMessage="by {user}"
        values={{ user }}
      />
    </StreetMetaItem>
  )
}

export default StreetMetaAuthor
