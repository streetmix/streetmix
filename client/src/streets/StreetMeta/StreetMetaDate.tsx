import React from 'react'

import { useSelector } from '~/src/store/hooks'
import DateTimeRelative from '~/src/app/DateTimeRelative'
import Icon from '~/src/ui/Icon'
import { isOwnedByCurrentUser } from '../owner'
import StreetMetaItem from './StreetMetaItem'

// Milliseconds that must have elapsed before recent edits display timestamp
const TIME_LIMIT = 6000

function StreetMetaDate (): React.ReactElement | null {
  const updatedAt = useSelector(
    (state) => state.street.updatedAt ?? state.street.clientUpdatedAt
  )

  if (updatedAt === undefined) return null

  // In the user has just edited the street, don't display the
  // "seconds ago" timestamp.
  if (isOwnedByCurrentUser()) {
    const now = new Date().getTime()
    const updated = new Date(updatedAt).getTime()
    if (now - updated < TIME_LIMIT) return null
  }

  return (
    <StreetMetaItem icon={<Icon name="time" />}>
      {/* Wrap in <span> so phrases like "Today at 8:43 PM" preserve the space
          between text and <time> elements */}
      <span>
        <DateTimeRelative value={updatedAt} />
      </span>
    </StreetMetaItem>
  )
}

export default StreetMetaDate
