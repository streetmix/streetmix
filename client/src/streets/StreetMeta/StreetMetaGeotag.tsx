import React from 'react'
import { FormattedMessage, useIntl } from 'react-intl'

import { useSelector, useDispatch } from '~/src/store/hooks'
import { showDialog } from '~/src/store/slices/dialogs'
import Icon from '~/src/ui/Icon'
import StreetMetaItem from './StreetMetaItem'

import type { StreetLocation } from '@streetmix/types'

function StreetMetaGeotag (): React.ReactElement | null {
  const street = useSelector((state) => state.street)
  const editable = useSelector(
    (state) => !state.app.readOnly && state.flags.GEOTAG.value
  )
  const dispatch = useDispatch()
  const intl = useIntl()

  // Render nothing if there is no street location, and geolocation is not enabled
  if (!editable && !street.location) return null

  function handleClickGeotag (event: React.MouseEvent): void {
    dispatch(showDialog('GEOTAG'))
  }

  function getGeotagText (
    location: StreetLocation
  ): string | React.ReactElement {
    const { hierarchy } = location
    const unknownLabel = (
      <FormattedMessage
        id="dialogs.geotag.unknown-location"
        defaultMessage="Unknown location"
      />
    )

    let text =
      hierarchy.locality ?? hierarchy.region ?? hierarchy.neighbourhood

    if (text !== undefined && hierarchy.country !== undefined) {
      text = text + ', ' + hierarchy.country
    }

    return text ?? unknownLabel
  }

  // Determine what text label to render
  const geotagText =
    street.location !== null
      ? (
          getGeotagText(street.location)
        )
      : (
        <FormattedMessage
          id="dialogs.geotag.add-location"
          defaultMessage="Add location"
        />
        )

  const tooltip = intl.formatMessage({
    id: 'tooltip.geotag',
    defaultMessage: 'Change location'
  })

  return (
    <StreetMetaItem
      isEditable={editable}
      tooltip={tooltip}
      onClick={handleClickGeotag}
      icon={<Icon name="location" />}
    >
      <span className="underline">{geotagText}</span>
    </StreetMetaItem>
  )
}

export default StreetMetaGeotag
