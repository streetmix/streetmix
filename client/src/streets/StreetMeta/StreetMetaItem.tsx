import React, { forwardRef } from 'react'

import Tooltip from '../../ui/Tooltip'
import './StreetMetaItem.css'

import type { TippyProps } from '@tippyjs/react'

interface StreetMetaItemProps {
  className?: string
  isEditable?: boolean // Might need a better name; isInteractive?
  tooltip?: string
  sublabel?: string
  onClick?: (event: React.MouseEvent) => void
  icon?: React.ReactNode
  children: React.ReactNode

  // We needed to type this for some special logic, below
  'data-state'?: string
}

const StreetMetaItem = forwardRef(
  (
    {
      className = '',
      isEditable = false,
      tooltip,
      sublabel,
      onClick = () => {},
      icon,
      children,
      // extra props are spread on child <button>, if present
      // This is needed for Radix UI's dropdown menu implementation.
      ...restProps
    }: StreetMetaItemProps,
    // Forwards a ref to child <button>, if present
    ref: React.Ref<HTMLButtonElement>
  ) => {
    function handleClick (event: React.MouseEvent): void {
      event.preventDefault()
      onClick(event)
    }

    const classNames = ['street-meta-item']
    if (!isEditable) {
      classNames.push('street-meta-item-plain')
    }
    if (className) {
      classNames.push(className)
    }

    const content = (
      <>
        {icon !== undefined && <div className="street-meta-icon">{icon}</div>}
        <div className="street-meta-content">{children}</div>
      </>
    )

    // A workaround for when this is a child of Radix UI's dropdown menu
    // When the dropdown opens, always hide the tooltip.
    // Unfortunately once the tooltip becomes controlled via `visible`,
    // it apparently longer acts in an uncontrolled manner. (maybe we can fix
    // this by setting uncontrolled props back, but what are they?)
    const maybeTooltipProps: Partial<TippyProps> = {}
    // Only act if this prop exists
    if (restProps['data-state'] !== undefined) {
      if (restProps['data-state'] === 'open') {
        maybeTooltipProps.visible = false
      } else {
        delete maybeTooltipProps.visible
      }
    }

    if (isEditable) {
      return (
        <div className={classNames.join(' ')}>
          <Tooltip
            label={tooltip}
            sublabel={sublabel}
            placement="bottom"
            {...maybeTooltipProps}
          >
            <button
              onClick={handleClick}
              aria-label={tooltip}
              ref={ref}
              {...restProps}
            >
              {content}
            </button>
          </Tooltip>
        </div>
      )
    }

    return <div className={classNames.join(' ')}>{content}</div>
  }
)

StreetMetaItem.displayName = 'StreetMetaItem'

export default StreetMetaItem
