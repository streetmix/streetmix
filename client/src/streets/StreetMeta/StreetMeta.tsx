import React from 'react'
import { useSelector } from '../../store/hooks'
import StreetMetaWidth from './StreetMetaWidth'
import StreetMetaAuthor from './StreetMetaAuthor'
import StreetMetaDate from './StreetMetaDate'
import StreetMetaGeotag from './StreetMetaGeotag'
import StreetMetaAnalytics from './StreetMetaAnalytics'
import './StreetMeta.scss'

function StreetMeta (): React.ReactElement {
  const enableAnalytics = useSelector(
    (state) => state.flags.ANALYTICS?.value ?? false
  )

  return (
    <div className="street-meta">
      <StreetMetaWidth />
      {enableAnalytics && <StreetMetaAnalytics />}
      <StreetMetaGeotag />
      <StreetMetaAuthor />
      <StreetMetaDate />
    </div>
  )
}

export default StreetMeta
