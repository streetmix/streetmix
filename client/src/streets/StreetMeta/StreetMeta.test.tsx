import React from 'react'
import { waitFor } from '@testing-library/react'

import { render } from '~/test/helpers/render'
import StreetMeta from './StreetMeta'

describe('StreetMeta', () => {
  it('renders without crashing', async () => {
    const { asFragment } = render(<StreetMeta />)

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot()
    })
  })
})
