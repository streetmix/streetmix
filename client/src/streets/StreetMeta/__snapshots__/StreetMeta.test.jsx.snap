// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StreetMeta > renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="street-meta"
  >
    <div
      class="street-meta-item street-width street-width-editable"
    >
      <button>
        <div
          class="street-meta-icon"
        >
          <svg />
        </div>
        <div
          class="street-meta-content"
        >
          <span
            class="underline"
          >
            0 m width
          </span>
          <svg
            class="menu-carat-down"
          />
        </div>
      </button>
    </div>
    <div
      class="street-meta-item"
    >
      <button>
        <div
          class="street-meta-icon"
        >
          <svg />
        </div>
        <div
          class="street-meta-content"
        >
          <span
            class="underline"
          >
            Add location
          </span>
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
